packages/next/src/server/profiler.ts

```diff
@@ -0,0 +1,102 @@
+import type { NodeNextRequest, NodeNextResponse } from './base-http/node'
+import type { Session } from 'inspector'
+
+let session: Session | null = null
+const PROFILER_ENABLED = !!process.env.NEXTJS_PROFILER_ENABLED
+const SAMPLING_INTERVAL = process.env.NEXTJS_PROFILER_INTERVAL
+  ? parseInt(process.env.NEXTJS_PROFILER_INTERVAL, 10)
+  : 100
+
+const promisify = <T>(
+  fn: (cb: (err: Error | null, result?: T) => void) => void
+) =>
+  new Promise<T>((resolve, reject) => {
+    fn((err, result) => {
+      err ? reject(err) : resolve(result!)
+    })
+  })
+
+async function maybeStopProfiler() {
+  if (!session) {
+    return
+  }
+
+  const { profile } = await promisify<{
+    profile: any
+  }>((cb) => session!.post('Profiler.stop', cb))
+  await promisify((cb) => session!.post('Profiler.disable', cb))
+
+  // not sure of the cost of keeping a session around, but it's best to clean up
+  session!.disconnect()
+  session = null
+
+  return JSON.stringify(profile)
+}
+
+function patchResponseObject(res: NodeNextResponse) {
+  const originalEnd = res.originalResponse.end.bind(res.originalResponse)
+  res.originalResponse.setHeader('x-nextjs-trace', '1')
+
+  // @ts-ignore
+  res.originalResponse.write = () => {}
+
+  res.originalResponse.setHeader = () => {}
+
+  // @ts-ignore
+  res.originalResponse.end = () => {
+    maybeStopProfiler().then((result) => {
+      originalEnd(result)
+    })
+  }
+}
+
+async function startProfiler() {
+  if (session) {
+    return
+  }
+
+  const inspector = await import('inspector')
+
+  session = new inspector.Session()
+  session.connect()
+
+  await promisify((cb) =>
+    session!.post(
+      'Profiler.setSamplingInterval',
+      {
+        interval: SAMPLING_INTERVAL,
+      },
+      cb
+    )
+  )
+  await promisify((cb) => session!.post('Profiler.enable', cb))
+  await promisify((cb) => session!.post('Profiler.start', cb))
+
+  setTimeout(() => {
+    maybeStopProfiler()
+  }, 5000)
+}
+
+export async function maybeStartProfiler() {
+  if (PROFILER_ENABLED) {
+    await startProfiler()
+  }
+}
+
+export async function maybeStartProfilingRequest(
+  req: NodeNextRequest,
+  res: NodeNextResponse
+) {
+  if (!PROFILER_ENABLED) {
+    return
+  }
```

Comment:
Having a branch (`if (!PROFILER_ENABLED)`) for every incoming request could be slower (the engine might have trouble optimizing this).

The creator of [Bluebird](https://github.com/petkaantonov/bluebird/) (Bluebird is one of the fastest JavaScript Promise implementations to this day, even comparable to the V8 native) has introduced an approach in [his blog post *Three JavaScript performance fundamentals that make Bluebird fast*](https://www.reaktor.com/blog/javascript-performance-fundamentals-make-bluebird-fast/) called "Use no-op functions and overwrite them lazily to implement costly optional features", here I quote:

> Checking if the monitoring feature is enabled before making the call to the hook function would be better than just making the call regardless of if the feature is enabled. However, the cost can actually be completely eliminated for users that have not enabled the feature thanks to inline caches and function inlining. This can be done by making the initial hook method just an empty function.
>
> Now, if the user doesn’t enable the monitoring feature, the optimizer sees that the function call doesn’t do anything and eliminates it. So the call to the hook method inside the constructor effectively doesn’t exist.
>
> To make the feature actually work, the act of enabling the feature must overwrite all the related no-op functions with the actual implementations.